async-support/src/lib.rs

```diff
@@ -0,0 +1,99 @@
+#![feature(generator_trait)]
+#![no_std]
+
+pub use core::*;
+
+pub mod future {
+    pub use core::future::Future;
+    use core::ops::Generator;
+    use core::pin::Pin;
+    use core::task::Poll;
+
+    pub fn poll_with_tls_context<F>(f: Pin<&mut F>) -> Poll<F::Output>
+    where
+        F: Future,
+    {
+        crate::executor::poll(f)
+    }
+
+    pub fn from_generator<G: Generator<Yield = ()>>(
+        generator: G,
+    ) -> impl Future<Output = G::Return> {
+        crate::executor::from_generator(generator)
+    }
+}
+
+pub mod executor {
+    use core::future::Future;
+    use core::ops::Generator;
+    use core::ops::GeneratorState;
+    use core::pin::Pin;
+    use core::ptr;
+    use core::task::Context;
+    use core::task::Poll;
+    use core::task::RawWaker;
+    use core::task::RawWakerVTable;
+    use core::task::Waker;
+
+    static DUMMY_WAKER_VTABLE: RawWakerVTable =
+        RawWakerVTable::new(clone, do_nothing, do_nothing, do_nothing);
+
+    extern "Rust" {
+        #[link_name = "libtock::syscalls::yieldk"]
+        fn yieldk();
+    }
```

Comment:
Why isn't this directly importing the function from libtock?

Reply:
This would either add a cyclic dependency or the `syscall` module of `libtock` would have to be moved to the `async-support` crate. I consider the `async-support` crate being a temporary solution that might get replaced by a `lang_item` or something else.

Reply:
In this case, it seems to me that the correct dependency graph is to introduce a separate `syscall` library (name TBD), on which the other two each depend. By extracting it from `libtock` there is no cycle. And libtock-rs would re-export the syscalls publicly.

Alternatively, why not put everything together in libtock-rs?

Reply:
I usually would agree that there should be another syscall library if this was a natural dependency. In this case, it's just a workaround for the temporary situation that the Rust team did not decide yet how the `poll_with_tls_context` and `from_generator` are to be implemented.

I am not strictly against the new crate but it adds even more churn to the build system and before I do it I would like to be sure that this is the preferred way for everyone. What are other people's opinions?

Regarding your question: To my best knowledge you cannot add a new item to the `core` library without creating a new crate.

Reply:
> I usually would agree that there should be another syscall library if this was a natural dependency. In this case, it's just a workaround for the temporary situation that the Rust team did not decide yet how the `poll_with_tls_context` and `from_generator` are to be implemented.

I'm not familiar with it, but do you mean that `poll_with_tls_context` and `from_generator` are unstable? Given that async/await is quite new in Rust, it's adding to my concerns of potential unstability of this new API, and (as mentioned in my other comments) to the importance of leaving the choice to use the previous API, until it's clear that the future-based API is stable enough for libtock-rs.

> Regarding your question: To my best knowledge you cannot add a new item to the `core` library without creating a new crate.

Makes sense.

Reply:
It means that the way `poll_with_tls_context` and `from_generator` are implemented will change. I expect it to transition to some kind of `lang_item` or attribute, so that you can remove the `async-support` and write the following:

```rust
#[poll_wth_tls_context]
pub fn poll_with_tls_context<F: Future>(f: Pin<&mut F>) -> Poll<F::Output> { ... }
```

This is an internal detail of `libtock-rs` and does not effect its API.